atef/widgets/archive_viewer.py

```diff
@@ -0,0 +1,622 @@
+    ...
+
+
+class ArchiverViewerWidget(DesignerDisplay, QWidget):
+    """
+    Archiver time plot viewer
+    """
+    filename: ClassVar[str] = 'archive_viewer_widget.ui'
+
+    time_plot: PyDMArchiverTimePlot
+    button_month: QtWidgets.QPushButton
+    button_week: QtWidgets.QPushButton
+    button_day: QtWidgets.QPushButton
+    input_field: QtWidgets.QLineEdit
+    curve_list: QtWidgets.QTableView
+    redraw_button: QtWidgets.QPushButton
+
+    def __init__(
+        self,
+        parent: Optional[QtWidgets.QWidget] = None,
+        pvs: List[str] = []
+    ) -> None:
+        super().__init__(parent=parent)
+
+        # set the PYDM_ARCHIVER_URL if not already set
+        if not os.environ.get('PYDM_ARCHIVER_URL'):
+            archiver_url = get_reachable_url(ARCHIVER_URLS)
+            if archiver_url is None:
+                raise ArchiverError('Cannot reach any archiver urls')
+            # need to set environment variable for archiver data plugin
+            logger.debug(f'setting archiver url to: {archiver_url}')
+            # pydm requires the port to be added
+            os.environ['PYDM_ARCHIVER_URL'] = archiver_url + ':17668'
+        else:
+            archiver_url = os.environ['PYDM_ARCHIVER_URL']
+            # ensure the port has been added for pydm
+            # this handling needs work, but should suffice for now
+            if not archiver_url.endswith(':17668'):
+                port_re = re.search(r'(:\d+)$', archiver_url)
+                if port_re and (port_re[0] != ':17668'):
+                    logger.warning(f'PYDM_ARCHIVER_URL ({archiver_url}) does '
+                                   'not end with port 17668, replacing port')
+                    archiver_url = archiver_url.replace(port_re[0], ':17668')
+                else:
+                    archiver_url += ':17668'
+                os.environ['PYDM_ARCHIVER_URL'] = archiver_url
+
+        # EpicsArchive wants a stripped down url
+        url_core = archiver_url.removeprefix('http://').removesuffix(':17668')
+        self.archapp = EpicsArchive(url_core)
+
+        self._pv_list = pvs
+        for pv in pvs:
+            self.model.add_signal(pv)
+
+        self._setup_ui()
+
+    def _setup_ui(self) -> None:
+        """
+        Initial set up for this widget.
+        - initializes PVModel
+        - assigns delegates to columns for data handling
+        - wire up various buttons
+        - adds pvs if provided at init
+        """
+        self.setWindowTitle('Epics PV Archive Viewer')
+
+        # set up table view for PV info
+        self.model = PVModel(parent=self)
+        self.curve_list.setModel(self.model)
+        horiz_header = self.curve_list.horizontalHeader()
+        horiz_header.setSectionResizeMode(horiz_header.ResizeToContents)
+
+        # set up delegates
+        # Color picker delegate
+        self.colorDelegate = ColorDelegate()
+        self.curve_list.setItemDelegateForColumn(2, self.colorDelegate)
```

Comment:
I wish I had picked up on these delegates when I was taking over `pmps-ui`, it's a much cleaner way to handle these tables

Reply:
It took me longer than I wanted to sort them out, but they do turn out quite nicely once they're working 👍